subset-routing-test/src/main/java/com/hazelcast/jet/tests/client/subset/routing/SubsetRoutingTest.java

```diff
@@ -0,0 +1,218 @@
+                            effectiveMemberList.size() == 1 && activeConnections == 1,
+                            new ConnectionsContext(activeConnections, effectiveMemberList));
+                })
+                .onConditionPass((context) -> {
+                    UUID newConnectedMemberUuid = context.effectiveMemberList
+                            .iterator().next().getUuid();
+                    if (!newConnectedMemberUuid.equals(CONNECTED_MEMBER_UUID.get())) {
+                        logger.info("Client is connected to a new member: "
+                                + newConnectedMemberUuid + " previously: " + CONNECTED_MEMBER_UUID.get());
+                    }
+                    CONNECTED_MEMBER_UUID.set(newConnectedMemberUuid);
+                })
+                .onConditionNotPassedYet(
+                        (context) -> logger.info("Waiting for expected connection... Currently,"
+                                + " active connections: "
+                                + context.activeConnections + " effective members: "
+                                + membersToString(context.effectiveMemberList)))
+                .onTimeout(
+                        () -> {
+                            throw new IllegalStateException("Timeout during waiting for expected connection " +
+                                    "and effective members." +
+                                    " Active connections: " + getActiveConnections(clientInstance) +
+                                    " Effective members:" + membersToString(getEffectiveMemberList(clientInstance)) +
+                                    " Previously client was connected to " + CONNECTED_MEMBER_UUID.get() + " member");
+                        })
+                .execute();
+    }
+
+    private static String membersToString(Collection<Member> effectiveMemberList) {
+        return '[' + effectiveMemberList
+                .stream()
+                .map(Member::getUuid)
+                .map(Objects::toString)
+                .collect(Collectors.joining(",")) + ']';
+    }
+
+    private static int getActiveConnections(HazelcastClientInstanceImpl clientInstance) {
+        return clientInstance.getConnectionManager().getActiveConnections().size();
+    }
+
+    private static Collection<Member> getEffectiveMemberList(HazelcastClientInstanceImpl clientInstance) {
+        return clientInstance.getClientClusterService().getEffectiveMemberList();
+    }
+
+    @Override
+    protected void teardown(Throwable t) throws Exception {
+    }
+
+    public static HazelcastClientInstanceImpl getHazelcastClientInstanceImpl(HazelcastInstance hz) {
+        HazelcastClientInstanceImpl impl = null;
+        if (hz instanceof HazelcastClientProxy) {
+            impl = ((HazelcastClientProxy) hz).client;
+        } else if (hz instanceof HazelcastClientInstanceImpl) {
+            impl = (HazelcastClientInstanceImpl) hz;
+        }
```

Comment:
Pattern matching for `instanceof` can be used here instead. See https://rules.sonarsource.com/java/?search=Pattern%20Matching%20for%20%22instanceof%22

Reply:
I tried that initially by our checkstyles are not ready for this yet :sweat_smile: (complained about this) and I didn't want to spend time on fixing it.